NeosModLoader/Utility/EnumerableInjector.cs

```diff
@@ -0,0 +1,159 @@
+        private Action prefix = nothing;
+        private Func<TIn, bool> preItem = yes;
+        private Func<TIn, TOut> transformItem;
+
+        /// <summary>
+        /// Gets called when the wrapped enumeration returned the last item.
+        /// </summary>
+        public Action Postfix
+        {
+            get => postfix;
+            set => postfix = value ?? throw new ArgumentNullException(nameof(value), "Postfix can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item, with the transformed item, and whether it was passed through.
+        /// First thing to be called after execution returns to the enumerator after a yield return.
+        /// </summary>
+        public Action<TIn, TOut, bool> PostItem
+        {
+            get => postItem;
+            set => postItem = value ?? throw new ArgumentNullException(nameof(value), "PostItem can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called before the enumeration returns the first item.
+        /// </summary>
+        public Action Prefix
+        {
+            get => prefix;
+            set => prefix = value ?? throw new ArgumentNullException(nameof(value), "Prefix can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item to determine whether it should be passed through.
+        /// </summary>
+        public Func<TIn, bool> PreItem
+        {
+            get => preItem;
+            set => preItem = value ?? throw new ArgumentNullException(nameof(value), "PreItem can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item to transform it, even if it won't be passed through.
+        /// </summary>
+        public Func<TIn, TOut> TransformItem
+        {
+            get => transformItem;
+            set => transformItem = value ?? throw new ArgumentNullException(nameof(value), "TransforItem can't be null!");
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EnumerableInjector{TIn, TOut}"/> class using the supplied input enumerable and transform function.
+        /// </summary>
+        /// <param name="enumerable">The enumerable to inject into and transform.</param>
+        /// <param name="transformItem">The transformation function.</param>
+        public EnumerableInjector(IEnumerable<TIn> enumerable, Func<TIn, TOut> transformItem)
+            : this(enumerable.GetEnumerator(), transformItem)
+        { }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EnumerableInjector{TIn, TOut}"/> class using the supplied input enumerator and transform function.
+        /// </summary>
+        /// <param name="enumerable">The enumerator to inject into and transform.</param>
+        /// <param name="transformItem">The transformation function.</param>
+        public EnumerableInjector(IEnumerator<TIn> enumerator, Func<TIn, TOut> transformItem)
```

Comment:
IDE analyzer doesn't like this:
> Non-nullable field 'transformItem' must contain a non-null value when exiting constructor. Consider declaring the field as nullable.

Reply:
I know, but it also didn't like me trying to define the parameter as non-nullable, saying I need to enable experimental language features. Which... I think also requires a higher target framework? at least all the other new language features do I think.